caikit_tgis_backend/managed_tgis_subprocess.py

```diff
@@ -0,0 +1,371 @@
+        """Terminate the subprocess, wait for it to exit"""
+        with self._mutex:
+            self._ensure_terminated()
+
+    def _ensure_terminated(self):
+        """Terminate the subprocess if it exists
+
+        _mutex should be locked before calling this
+        """
+        if self._tgis_state != _TGISState.STOPPED:
+            if self._tgis_proc:
+                self._tgis_proc.terminate()
+                self._tgis_proc = None
+                # disown the bootup thread (but keep the exception if one was captured)
+                self._bootup_thread = None
+            log.debug("_ensure_terminated setting the state to STOPPED")
+            self._tgis_state = _TGISState.STOPPED
+
+    def _launch(self):
+        """Launch the subprocess or restart if it exists
+
+        _mutex should be locked before calling
+        """
+        self._ensure_terminated()
+
+        # Launch TGIS
+        launch_cmd = " ".join(
+            [
+                "text-generation-launcher",
+                "--num-shard 1",
+                f"--model-name {self._model_path}",
+                f"--port {self._http_port}",
+            ]
+        )
+        log.debug2("Launching TGIS with command: [%s]", launch_cmd)
+        env = os.environ.copy()
+        env["GRPC_PORT"] = str(self._grpc_port)
+        # Long running process
+        # pylint: disable=consider-using-with
+        self._tgis_proc = subprocess.Popen(shlex.split(launch_cmd), env=env)
+        log.debug("_launch setting state to BOOTING")
+        self._tgis_state = _TGISState.BOOTING
+        # launch background thread to monitor the boot sequence
+        # save a handle to it to record bootup errors
+        self._bootup_thread = threading.Thread(target=self._monitor_bootup, daemon=True)
+        self._bootup_thread.start()
+
+    def _monitor_bootup(self):
+        """Ran in a background thread to complete the BOOTING -> READY transition
+
+        Polls the TGIS process until if finishes booting. The polling is guarded
+        with a timeout.
+        """
+        # Wait for the server to be ready
+        start_t = time.time()
+        while True:
+            time.sleep(self._bootup_poll_delay)
+
+            # if we are not still booting, exit
+            # another thread could have caused a state transition
+            if self._tgis_state != _TGISState.BOOTING:
+                return
+
+            # check if the process stopped/crashed
+            with self._mutex:
+                try:
+                    if self._tgis_proc.poll() is not None:
+                        if self._tgis_state != _TGISState.STOPPED:
+                            self._ensure_terminated()
+                        exc = RuntimeError(
+                            "TGIS failed to boot up with the model. See logs for details"
+                        )
+                        self._bootup_exc = exc
+                        error("<MTS11752287E>", exc)
+                except AttributeError:
```

Comment:
Is there a specific `AttributeError` that you plan to catch here? It's not immediately clear to me why this would happen

Reply:
The `AttributeError` would arise if `self._tgis_proc` is `None` in the `if self._tgis_proc.poll()` conditional, which could happen if the subprocess was terminated with the `_monitor_bootup()` thread running in the background.

I'll make that check more explicit though and remove the try-catch.